packages/go_router/lib/src/parser.dart

```diff
@@ -124,3 +128,13 @@ class GoRouteInformationParser extends RouteInformationParser<RouteMatchList> {
     );
   }
 }
+
+/// Pre-parsed [RouteInformation] that contains a [RouteMatchList].
+class PreParsedRouteInformation extends RouteInformation {
+  /// Creates a [PreParsedRouteInformation].
+  PreParsedRouteInformation(
+      {super.location, super.state, required this.matchlist});
```

Comment:
This may cause issue with state restoration, but I think we should update the router to be able to handle RouteInformation subclass.

We should at least check the location generated from matchlist is the same as the input location though

Reply:
Edit: We can we create a special class to insert both extra and matchlist into RouteInformation.state? and the parser can be smart enough to read the class and separate this two before it proceed to parsing. That way we don't need to worry about subclass issue

Reply:
I'm working on state restoration now, and I think we perhaps may have to handle that in a separate StatefulWidget (added in RouteBuilder), since we probably need to serialize the location (graph) of the matchlists as strings somehow, and then parse them upon restoration. But we'll see.

Reply:
serialize the location (graph) of the matchlists as strings somehow

yes, think so. but I don't think we need  a separate StatefulWidget as long as we are able to rebuild the widget with the RouteInformation